blobstore/presigned_url.go

```diff
@@ -225,3 +260,128 @@ func (bh *BlobHandler) HandleGetPresignedURLMultiObj(c echo.Context) error {
 	log.Info("HandleGetPresignedURLMultiObj: Successfully generated presigned URL for prefix:", prefix)
 	return c.JSON(http.StatusOK, string(href))
 }
+
+func (bh *BlobHandler) HandleGenerateDownloadScript(c echo.Context) error {
+	prefix := c.QueryParam("prefix")
+	bucket := c.QueryParam("bucket")
+	if prefix == "" || bucket == "" {
+		errMsg := fmt.Errorf("`prefix` and `bucket` query params are required")
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
+	prefix = strings.TrimSuffix(prefix, "/")
+	s3Ctrl, err := bh.GetController(bucket)
+	if err != nil {
+		errMsg := fmt.Errorf("error getting controller for bucket %s: %s", bucket, err)
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusUnprocessableEntity, errMsg.Error())
+	}
+
+	//list objects within the prefix and test if empty
+	response, err := s3Ctrl.GetList(bucket, prefix, false)
+	if err != nil {
+		errMsg := fmt.Errorf("error listing objects in bucket %s with prefix %s: %s", bucket, prefix, err)
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusInternalServerError, errMsg.Error())
+	}
+	if len(response.Contents) == 0 {
+		errMsg := fmt.Errorf("prefix %s is empty or does not exist", prefix)
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusBadRequest, errMsg.Error())
+	}
+	size, _, err := bh.GetSize(response)
+	if err != nil {
+		errMsg := fmt.Errorf("error retrieving size of prefix: %s", err.Error())
+		log.Error(errMsg.Error())
+		return c.JSON(http.StatusInternalServerError, errMsg.Error())
+	}
+	var downloadLimit int
+	downloadLimit, err = strconv.Atoi(os.Getenv("SCRIPT_DOWNLOAD_SIZE_LIMIT"))
+	if err != nil {
+		log.Debugf("size download limit defaulted to %v", DEFAULT_SCRIPT_DOWNLOAD_SIZE_LIMIT)
+		downloadLimit = DEFAULT_SCRIPT_DOWNLOAD_SIZE_LIMIT
+	}
+	limit := uint64(1024 * 1024 * 1024 * downloadLimit)
+	if size > limit {
+		errMsg := fmt.Errorf("request entity is larger than %v GB, current prefix size is: %v GB", float64(limit)/(1024*1024*1024), float64(size)/(1024*1024*1024))
```

Comment:
See comment above about extra math. Not that important here, but in general it is best practice to not back calculate a known value to avoid rounding errors